### PR TITLE
core: migrate Region.ops to _BlockOps

### PR DESCRIPTION
**Pull request by superlopuh**
Users will have to explicitly construct a list if they want to.

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Why would we want to do this? Copying all ops seems asymptotically costly? Should ops not jus point to bb.ops?

**Review by math-fehr — APPROVED** [review]
:+1:, we want to expose the iterator API, but don't necessarily want to consume the iterator, so this makes more sense!
I would however return `Sequence[Operation]` rather than `_BlockOps`, as `_BlockOps` is an implementation detail, and we shouldn't rely on that.

**Comment by tobiasgrosser** [comment]
I am still unsure why we now need to create lists all over the place?

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -1244,7 +1244,7 @@ def __init__(self, ops: List[Operation] | Region):
 
     @property
     def ops(self) -> List[Operation]:
```
Comment (math-fehr) [inline xdsl/dialects/builtin.py:1246]: Actually yes, this should return the iterator instead.

**Comment by math-fehr** [comment]
So we create lists in tests there, but indeed there was one case where we don't care about returning a list!

File: xdsl/dialects/builtin.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/builtin.py:1246]: Upcoming PR :)

**Comment by superlopuh** [comment]
There's unfortunately no interface that's `Iterable` + `__len__(self)`, as far as I know, so I can't just replace `_BlockOps` without having to migrate some more code. I'm happy to do that if you think it's worth it.

File: xdsl/dialects/builtin.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/dialects/builtin.py:1246]: Okay then fine!

**Comment by math-fehr** [comment]
Oh sorry, I'm always confused with these names! Can we make it just `Iterable`?

**Comment by superlopuh** [comment]
The `ops` property is a helper that seems to mostly be used in tests, but at least with `list(region.ops)` it's now explicit to users that it's potentially expensive, and that it's worth thinking of an alternative. Currently code is silently expensive, this API change at least makes it uglier to do the expensive thing.

**Comment by superlopuh** [comment]
No, that's what I mean, `Sequence` is `Iterable` with length and indexing, but there's nothing inbetween that's `Sequence` without indexing

**Comment by math-fehr** [comment]
Oh I see now...
In another related note, how much do we need `len`? Is there a use case somewhere of using it?
I feel that if you want the len, you should manually cast it to a list, as you should know it's an expensive computation?

Otherwise, we should probably do that in another PR, so I'm good here!

**Comment by superlopuh** [comment]
It's O(n) time but that's not as expensive as creating a list, so still seems useful to me. I'd guess that there's not much use doing it in production code, as oposed to tests, so doesn't seem like a big footgun. Let's see if we run into this problem

**Comment by math-fehr** [comment]
Yeah okay, if you think that's worth we can keep it!
It's just that in terms of API, we are now exposing an `_BlockOps` that is a private class, so users won't necessarily know what's going on.
I'm wondering if we should rename it `BlockOpsIterator`? So we can expose it to users and they will understand what is going on?

**Comment by superlopuh** [comment]
Happy to do that, I don't think there's any risk in it being public. Can do it in this PR, or another one, up to you

**Comment by math-fehr** [comment]
Yeah I'm fine with both!
Though having small PRs is nice, sometimes it's better to not introduce too many problems between PRs.
But here I'm fine, as this won't really affect people now!

**Comment by math-fehr** [comment]
So yeah, I'm fine with the PR now!

**Review by webmiche — APPROVED** [review]
LGTM. Making the expensive thing harder to do usually feels like a good design choice to me :)

**Review by georgebisbas — APPROVED** [review]

**Comment by superlopuh** [comment]
Merged these changes in #795, closing

**Pull request closed without merging**